comet/manager.py

```diff
@@ -443,84 +420,155 @@ def _check_result(self, result, endpoint):
         -------
-        dict
-            The last config or state of requested type that was registered. Returns `None` if
-            requested state not found.
+        State
+            The last config or state of requested type that was registered. Returns
+            `None` if requested state not found.
+
+        Raises
+        ------
+        ManagerError
+            If an error in the dataset tree is encountered.
         """
-        if dataset_id is not None:
-            raise ManagerError(
-                "get_state: Not implemented for dataset_id other than None."
-            )
+        if dataset_id is None:
+            # return whatever was registered last of that type
+            if type is None:
+                return self.states[self.config_state]
+
+            states_of_type = list()
+            for state_id in self.states:
+                state = self.states[state_id]
+                if state.type == type:
+                    states_of_type.append(state)
+
+            if states_of_type:
+                states_of_type.sort(key=lambda s: self.state_reg_time[state_id])
+                return states_of_type[-1]
+            return None
+        else:
+            # traverse the tree towards the root to find a matching state type
+            while True:
+                dataset = self.get_dataset(dataset_id)
+                if type is None or type == dataset.state_type:
+                    return self._get_state(dataset.state_id)
+                if dataset.is_root == True:
+                    return None
+                dataset_id = dataset.base_dataset_id
+                if dataset_id is None:
+                    raise ManagerError(
+                        "Found a dataset that is not root nor has a base dataset ID: {}".format(
+                            dataset.to_dict()
+                        )
+                    )
+
+    def _get_state(self, state_id):
+        """
+        Get a state by ID.
 
-        if type is None:
-            return self.states[self.config_state]
+        If not known locally, the dataset broker is asked. Returns `None` if the state
+        is still unknown.
 
-        states_of_type = list()
-        for state_id in self.states:
-            state = self.states[state_id]
-            if state["type"] == type:
-                states_of_type.append(state)
+        Parameters
+        ----------
+        state_id : int
+            ID of the requested state.
 
-        if states_of_type:
-            states_of_type.sort(key=lambda s: self.state_reg_time[state_id])
-            return states_of_type[-1]
-        return None
+        Returns
+        -------
+        dict or None
+            The requested state.
+        """
+        try:
+            return self.states[state_id]
+        except KeyError:
+            try:
+                reply = self._send(REQUEST_STATE, {"id": state_id}, "post")
+                self.states[state_id] = State.from_dict(reply["state"])
```

Comment:
I think this is a little dangerous. When you create the `State` like this it's going to rehash the internal data to produce the `State.id` property. It's very possible that we won't have the exact same JSON serialisation and hashing in both kotekan and in comet, and so it could give different hashes and thus it might be that `self.state[state_id].id != state_id`, which would be bad.

I would recommend adding an optional field to the constructor of `State` that explicitly sets the ID (or if not set does the hashing itself). Alternatively you could set `self.states[state_id]._id = state_id` in here.